File: src/javascript/xml_parser.js
class Utilities {
  static hasValue(x) {
    return x !== null && x !== undefined;
  }

  static withDefault(x, xDefault) {
    return Utilities.hasValue(x) ? x : xDefault;
  }
}

class XmlTag {
  tagString;
  firstIndex;
  secondIndex;
  attributes;

  constructor(
    tagString,
    firstIndex,
    secondIndex
  ) {
    this.tagString = tagString;
    this.firstIndex = firstIndex;
    this.secondIndex = secondIndex;
    this.attributes = this.getAtributes();
  }
  
  static getNext(xmlString, searchStartIndex) {
    const xmlStringLength = xmlString.length;
    let firstIndex = -1;
    let secondIndex = -1;
    let index = searchStartIndex;
    while (secondIndex < 0 && index < xmlStringLength) {
      const c = xmlString.charAt(index);
      if (c === '<') {
        if (firstIndex > 0) {
          throw new Error('Unexpected < was encountered.');
        }
        firstIndex = index;
      } else if (c === '>') {
        if (firstIndex < 0) {
          throw new Error('Unexpected > was encountered.');
        }
        secondIndex = index;
      }
      ++index;
    }
    if (secondIndex < 0) {
      return null;
    }
    return new XmlTag(
      xmlString.substring(firstIndex, secondIndex + 1),
      firstIndex,
      secondIndex
    );
  }

  getAtributes() {
    const keyValues = {};
    this.getNameAndAtributes()
      .slice(1)
      .map((x) => x.trim().split('='))
      .forEach((x) => keyValues[x[0].trim()] = x[1].trim().replaceAll('"', ''))
    ;
    return keyValues;
  }

  getName() {
    return this.getNameAndAtributes()[0];
  }

  getNameAndAtributes() {
    if (!this.isValid()) {
      throw new Error('Invalid Tag was encountered.');
    }
    return this.tagString.substring(this.isEndTag() ? 2 : 1, this.tagString.length - 1).split(' ');
  }

  isEndTag() {
    if (!this.isValid()) {
      throw new Error('Invalid Tag was encountered.');
    }
    return this.tagString.startsWith('</');
  }

  isValid() {
    return (
      Utilities.hasValue(this.tagString) &&
      this.tagString.startsWith('<') &&
      this.tagString.charAt(this.tagString.length - 1) === '>' &&
      Number.isInteger(this.firstIndex) &&
      Number.isInteger(this.secondIndex)
    );
  }
}

class XmlNode {
  parent;
  name;
  attributes;
  children;
  value;

  constructor(parent, name, attributes, children, value) {
    this.parent = parent;
    this.name = name;
    this.attributes = attributes;
    this.children = Utilities.withDefault(children, []);
    this.value = value;
    if (Utilities.hasValue(parent)) {
      parent.children.push(this);
    }
  }
  
  static create(parent, name, attributes) {
    return new XmlNode(parent, name, attributes, [], null);
  }
  
  isLeaf() {
    return this.children.length === 0;
  }

  setValue(value) {
    const hasValue = Utilities.hasValue(value);
    if (hasValue && !this.isLeaf()) {
      throw new Error('Value cannot be assigned to a non-leaf node.');
    }
    this.value = hasValue ? value.toString().replaceAll('\\n', '').trim() : null;
  }

  static removeParentsAndNulls(node) {
    if (!Utilities.hasValue(node)) {
      return;
    }
    node.parent = undefined;
    for (const child of node.children) {
      XmlNode.removeParentsAndNulls(child);
    }
    if (node.children.length === 0) {
      node.children = undefined;
    }
    if (!Utilities.hasValue(node.value)) {
      node.value = undefined;
    }
    if (!Utilities.hasValue(node.attributes) || Object.keys(node.attributes).length === 0) {
      node.attributes = undefined;
    }
  }
}

export class XmlParser {
  static parse(xmlString, extractAllTagsFirst) {
    if (typeof xmlString !== 'string') {
      throw new Error('Invalid XML-string was encountered.');
    }
    const xmlNodeRoot = XmlNode.create(null, null);
    if (extractAllTagsFirst) {
      // Method 1: more readable, more separation of concerns, but with more space complexity
      const tags = XmlParser.getTags(xmlString);
      XmlParser.addNodes(xmlString, tags, xmlNodeRoot);
    } else {
      // Method 2: less readable, less separation of concerns, but with less space complexity
      XmlParser.extractTagsAndAddNodes(xmlString, xmlNodeRoot);
    }
    const root = xmlNodeRoot.children[0];
    return root;
  }

  static extractTagsAndAddNodes(xmlString, parent) {
    const xmlStringLength = xmlString.length;
    let index = 0;
    let lastNode = parent;
    let previousTag = null;
    while (index < xmlStringLength) {
      const tag = XmlTag.getNext(xmlString, index);
      if (!Utilities.hasValue(tag)) {
        return;
      } else {
        if (!tag.isEndTag()) { // <TAG>
          const parent = (
            Utilities.hasValue(previousTag) &&
            previousTag.isEndTag() &&
            Utilities.hasValue(lastNode)
            ) ? lastNode.parent : lastNode;
          lastNode = XmlNode.create(parent, tag.getName(), tag.getAtributes());
        } else { // </TAG>
          if (!Utilities.hasValue(previousTag)) {
            throw new Error('Unexpected </TAG> was encountered.');
          }
          if (!previousTag.isEndTag()) {
            if (previousTag.getName() === tag.getName()) {
              const value = xmlString.substring(
                previousTag.secondIndex + 1,
                tag.firstIndex
              );
              lastNode.setValue(value);
            } else {
              throw new Error('Unpaired tag names were encountered: ' +
              tag.getName() + ' vs. ' + previousTag.getName())
            }
          }
        }
        previousTag = tag;
        index = tag.secondIndex + 1;
      }
    }
  }

  static getTags(xmlString) {
    const xmlStringLength = xmlString.length;
    const tags = [];
    let index = 0;
    while (index < xmlStringLength) {
      const tag = XmlTag.getNext(xmlString, index);
      if (Utilities.hasValue(tag)) {
        tags.push(tag);
        index = tag.secondIndex + 1;
      } else {
        index = xmlStringLength;
      }
    }
    return tags;
  }

  static addNodes(xmlString, tags, parent) {
    let lastNode = parent;
    let previousTag = null;
    for (const tag of tags) {
      if (!tag.isEndTag()) { // <TAG>
        const parent = (
          Utilities.hasValue(previousTag) &&
          previousTag.isEndTag() &&
          Utilities.hasValue(lastNode)
          ) ? lastNode.parent : lastNode;
        lastNode = XmlNode.create(parent, tag.getName(), tag.getAtributes());
      } else { // </TAG>
        if (!Utilities.hasValue(previousTag)) {
          throw new Error('Unexpected </TAG> was encountered.');
        }
        if (!previousTag.isEndTag()) {
          if (previousTag.getName() === tag.getName()) {
            const value = xmlString.substring(
              previousTag.secondIndex + 1,
              tag.firstIndex
            );
            lastNode.setValue(value);
          } else {
            throw new Error('Unpaired tag names were encountered: ' +
            tag.getName() + ' vs. ' + previousTag.getName())
          }
        }
      }
      previousTag = tag;
    }
  }

  static removeParentsAndNulls(node) {
    return XmlNode.removeParentsAndNulls(node);
  }
}

File: src/python/xml_parser.py
class XmlTag:
  def __init__(self, tag_string, first_index, second_index):
    self.tag_string = tag_string
    self.first_index = first_index
    self.second_index = second_index
    self.attributes = self.get_attributes()
  
  @staticmethod
  def get_next(xml_string, search_string_index):
    xml_stringLength = len(xml_string)
    first_index = -1
    second_index = -1
    index = search_string_index
    while (second_index < 0 and index < xml_stringLength):
      c = xml_string[index]
      if (c == '<'):
        if (first_index > 0):
          raise Exception('Unexpected < was encountered.')
        first_index = index
      elif (c == '>'):
        if (first_index < 0):
          raise Exception('Unexpected > was encountered.')
        second_index = index
      index += 1
    if (second_index < 0):
      return None
    return XmlTag(
      xml_string[first_index : second_index + 1],
      first_index,
      second_index
    )

  def get_attributes(self):
    key_values = {}
    attributes = self.get_name_and_attributes()[1:]
    if (len(attributes) == 0):
      return None
    for attribute in attributes:
      key_value = attribute.strip().split('=')
      key_values[key_value[0].strip()] = key_value[1].strip().replace('"', '')
    return key_values

  def get_name(self):
    return self.get_name_and_attributes()[0]
 
  def get_name_and_attributes(self):
    if (not self.is_valid()):
      raise Exception('Invalid Tag was encountered.')
    return self.tag_string[2 if self.is_end_tag() else 1: len(self.tag_string) - 1].split(' ')
 
  def is_end_tag(self):
    if (not self.is_valid()):
      raise Exception('Invalid Tag was encountered.')
    return self.tag_string.startswith('</')

  def is_valid(self):
    return self.tag_string != None and self.tag_string.startswith('<') and self.tag_string[len(self.tag_string) - 1] == '>' and isinstance(self.first_index, int) and isinstance(self.second_index, int) 


class XmlNode:
  def __init__(self, parent, name, attributes, children, value):
    self.parent = parent
    self.name = name
    self.attributes = attributes
    self.children = children if children != None else []
    self.value = value
    if (parent != None):
      parent.children.append(self)
  
  @staticmethod
  def create(parent, name, attributes):
    return XmlNode(parent, name, attributes, [], None)
  
  def isLeaf(self):
    return len(self.children) == 0

  def set_value(self, value):
    if (value == None):
      self.value = None
    else:
      if (not self.isLeaf()):
        raise Exception('Value cannot be assigned to a non-leaf node.')
      self.value = value.replace('\\n', '').strip()


class XmlParser:
  
  @staticmethod
  def parse(xml_string, extractAllTagsFirst):
    if (not isinstance(xml_string, str)):
      raise Exception('Invalid XML-string was encountered.')
    xmlNodeRoot = XmlNode.create(None, None, None)
    if (extractAllTagsFirst):
      # Method 1: more readable, more separation of concerns, but with more space complexity
      tags = XmlParser.getTags(xml_string)
      XmlParser.addNodes(xml_string, tags, xmlNodeRoot)
    else:
      # Method 2: less readable, less separation of concerns, but with less space complexity
      XmlParser.extractTagsAndAddNodes(xml_string, xmlNodeRoot)
    root = xmlNodeRoot.children[0]
    return root

  @staticmethod
  def extractTagsAndAddNodes(xml_string, parent):
    xml_string_length = len(xml_string)
    index = 0
    last_node = parent
    previous_tag = None
    while (index < xml_string_length):
      tag = XmlTag.get_next(xml_string, index)
      if (tag == None):
        return
      else:
        if (not tag.is_end_tag()): # <TAG>
          last_node_is_parent = (
            previous_tag != None and
            previous_tag.is_end_tag() and
            last_node != None
            )
          parent = last_node.parent if last_node_is_parent else last_node
          last_node = XmlNode.create(parent, tag.get_name(), tag.get_attributes())
        else: # </TAG>
          if (previous_tag == None):
            raise Exception('Unexpected </TAG> was encountered.')
          isLeaf = (
            not previous_tag.is_end_tag() and
            previous_tag.get_name() == tag.get_name()
          )
          if (isLeaf):
            value = xml_string[previous_tag.second_index + 1 : tag.first_index]
            last_node.set_value(value)
        previous_tag = tag
        index = tag.second_index + 1

  @staticmethod
  def getTags(xml_string):
    xml_string_length = len(xml_string)
    tags = []
    index = 0
    while (index < xml_string_length):
      tag = XmlTag.get_next(xml_string, index)
      if (tag != None):
        tags.append(tag)
        index = tag.second_index + 1
      else:
        index = xml_string_length
    return tags

  @staticmethod
  def addNodes(xml_string, tags, parent):
    last_node = parent
    previous_tag = None
    for tag in tags:
      if (not tag.is_end_tag()): # <TAG>
        last_node_is_parent = (
          previous_tag != None and
          previous_tag.is_end_tag() and
          last_node != None
          )
        parent = last_node.parent if last_node_is_parent else last_node
        last_node = XmlNode.create(parent, tag.get_name(), tag.get_attributes())
      else: # </TAG>
        if (previous_tag == None):
          raise Exception('Unexpected </TAG> was encountered.')
        isLeaf = (
          not previous_tag.is_end_tag() and
          previous_tag.get_name() == tag.get_name()
        )
        if (isLeaf):
          value = xml_string[previous_tag.second_index + 1 : tag.first_index]
          last_node.set_value(value)
      previous_tag = tag
